### PR TITLE
Add implementation of inner CFG generation

### PR DESCRIPTION
**Pull request by bebidek**
At the moment it is incomplete and probably buggy, not ready for review/merge.

### CODE REVIEW

File: app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt
Comment (ignacy123) [inline app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt:109]: There's a [builder class](https://github.com/TCS-Compilers/compiler/blob/function-details-generator-generate-call/app/src/main/kotlin/compiler/intermediate_form/ControlFlowGraphBuilder.kt) performing very similar operations being implemented on a parallel branch (which will be merged soon). Just letting you know, as this logic (and the method below)  can probably be easily moved there.

**Comment by bebidek** [comment]
Probably not a final version (waiting for bugfix in tests), but I consider it ready for review.

**Comment by bebidek** [comment]
Now is passes all the tests, ready for review.

File: app/src/main/kotlin/compiler/intermediate_form/IntermediateFormTreeNode.kt
Comment (ignacy123) [inline app/src/main/kotlin/compiler/intermediate_form/IntermediateFormTreeNode.kt:8]: Is this convention used in the whole repository (`Unit` being a zero value). If not, then create a GitHub issues to unify this at some point.

File: app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt
Comment (ignacy123) [inline app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt:52]: Do we need to have that many nested functions? I think that most of these can easily be just a private function. This would decrease the indentation level and shorten the main function significantly.

**Review by ignacy123 — APPROVED** [review]
I didn't go too deeply into the details, I believe that the tests are exhaustive enough to catch (almost) all the possible mistakes.
Please merge this branch after the branch with tests and address all the comments before that. Also make sure that you are merging into the `master` branch

File: app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt:165]: I have no idea, but the decision was (somehow, probably in a distributed way...) made earlier.

File: app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt:130]: The problem is that is uses additional state (variable `last`) analogously to `createGraphForEachFunction`.

File: app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt:68]: Changed.

File: app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt:52]: Changed.

File: app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt:137]: Changed.

File: app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt:270]: Changed.

File: app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt:173]: Changed, but using Triple instead of listOf, as list would require all three items to have the same type.

File: app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/ControlFlow.kt:52]: I am afraid that all these nested functions use some of `createGraphForExpression`'s arguments. In order to move them outside we would have to pass them directly (but I consider this solution less readable) or make `ControlFlow` full featured class (with constructor parameters) instead of singleton object.

File: app/src/main/kotlin/compiler/intermediate_form/IntermediateFormTreeNode.kt
Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/IntermediateFormTreeNode.kt:8]: The only place (at the moment) was `createGraphForEachFunction`, which I changed.

Comment (bebidek) [reply app/src/main/kotlin/compiler/intermediate_form/IntermediateFormTreeNode.kt:8]: OK, turns out that I can't change that because tests for `createGraphForEachFunction` require the possibility of empty CFG for an expression. Nevertheless, at the moment this is the only place assuming anything about representation of unit.

**Pull request merged**
